File: src/main/scala/net/edmacdonald/jvmcli/history/history.scala
package net.edmacdonald.jvmcli.history

import net.edmacdonald.jvmcli.history.History.getBean
import net.edmacdonald.jvmcli.repository.CommandInvocationRepository
import org.aspectj.lang.annotation.{Aspect, Before, Pointcut}
import org.springframework.context.{ApplicationContext, ApplicationContextAware}
import org.springframework.stereotype.Component

@Aspect
@Component
class History extends ApplicationContextAware {

  @Pointcut("execution(public * picocli.CommandLine.execute(..))")
  private def commandExecution(): Unit = {}

  @Before("net.edmacdonald.jvmcli.history.History.commandExecution()")
  def log(): Unit = {
    var commandInvocationRepository: CommandInvocationRepository =
      getBean(classOf[CommandInvocationRepository])

    println(s"EFM: Before command execution: ${commandInvocationRepository}")
  }

  override def setApplicationContext(ctx: ApplicationContext): Unit = {
    History.context = ctx
  }
}

object History {
  @volatile private var context: ApplicationContext = _

  def getBean[T](clazz: Class[T]): T = {
    context.getBean(clazz)
  }
}

File: src/main/java/net/edmacdonald/jvmcli/command/history/History.java
package net.edmacdonald.jvmcli.command.history;

import net.edmacdonald.jvmcli.command.Main;
import net.edmacdonald.jvmcli.command.ParentCommand;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import picocli.CommandLine;

import java.util.List;

@Component
@CommandLine.Command(name = "history", mixinStandardHelpOptions = true, description = "Commands related to command history.")
public class History extends ParentCommand<History.HistorySubCommand> implements Main.MainSubCommand {

    @Autowired
    public History(List<History.HistorySubCommand> subCommands) {
        this.subCommands = subCommands;
    }

    public interface HistorySubCommand {
    }
}
